### PR TITLE
Remove flag --no-dry-run for arduino-cli >= 0.14.0

### PR DESCRIPTION
**Pull request by hel06492**
## Highlights from `CHANGELOG.md`

* See CHANGELOG.md for more

### CODE REVIEW

**Comment by jgfoster** [comment]
@hel06492, Thank you for this PR. Could you give us a bit of background on why it is desirable or what problem it solves? Next, why tie it to an arduino_ci version? If it is desirable, why not have it take effect immediately? Presumably people using the old version will already have the legacy behavior. Thanks!

**Comment by hel06492** [comment]
@jgfoster The `--dry-run` flag was removed from version 0.14.0 and as for version 0.19.0 arduino-cli no longer accepts the flag. It fails with the following error. I can't find the changelog, only [a forum post](https://forum.arduino.cc/t/dry-run-compile-flag-no-more-available-in-cli-0-14-0-and-optimization-flags/687004) about the change.
```
Compiling Blink.ino for arduino:avr:mega:cpu=atmega2560... 
Last command:  $  /usr/bin/arduino-cli --format json compile --fqbn arduino:avr:mega:cpu=atmega2560 --warnings all --dry-run /home/salaryman/Arduino/trial.errors/Blink/examples/Blink/Blink.ino
Error: unknown flag: --dry-run
```

The version check is for older versions which might give unexpected behavior without the `--dry-run` flag.

Please let me know if there are more things I need to fix/add for this pull request. Thanks.

**Comment by jgfoster** [comment]
@hel06492, I think some of my confusion is that the change refers to `arduino_ci` (Continuous Integration) while it appears that it should refer to `arduino_cli` (Command Line Interface). Could you update the title and other text in the PR and comments in the code to make that clarification? Thanks!

File: lib/arduino_ci/arduino_backend.rb [hunk 1]
```diff
@@ -235,5 +239,14 @@ def install_local_library(path)
       Host.symlink(src_path, destination_path)
       cpp_library
     end
+
+    private
+
+    def should_use_dry_run?
+      ret = capture_json("version")
+      version = ret[:json]["VersionString"]
+      match = /(\d)\.(\d+)\.(\d+)/.match(version)
```
Comment (ianfixes) [inline lib/arduino_ci/arduino_backend.rb:248]: Use `Gem::Version` instead of regex to do this comparison:
https://stackoverflow.com/a/3064161

**Review by jgfoster — CHANGES_REQUESTED** [review]

File: lib/arduino_ci/arduino_backend.rb (hunk 1 above)
Comment (hel06492) [reply lib/arduino_ci/arduino_backend.rb]: Done

File: CHANGELOG.md [hunk 2]
```diff
@@ -14,6 +14,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - Support for `dtostrf()`
 - Added a CI workflow to lint the code base
 - Added a CI workflow to check for spelling errors
+- Remove `--dry-run` flag for `arduino_ci` version `>= 0.14.0`
```
[the remark this answers is not in the document]
Comment (hel06492) [reply CHANGELOG.md]: Done

File: spec/arduino_backend_spec.rb [hunk 3]
```diff
@@ -103,5 +103,48 @@ def get_sketch(dir, file)
     it "Passes a simple INO sketch at #{sketch_path_ino}" do
       expect(backend.compile_sketch(sketch_path_ino, "arduino:avr:uno")).to be true
     end
+
+    context "--dry-run flags" do
+      before { allow(backend).to receive(:run_and_capture).and_call_original }
+
+      it "Uses --dry-run flag for arduino_cli version < 0.14.0" do
```
[the remark this answers is not in the document]
Comment (hel06492) [reply spec/arduino_backend_spec.rb]: Done

File: spec/arduino_backend_spec.rb [hunk 4]
```diff
@@ -103,5 +103,48 @@ def get_sketch(dir, file)
...
+          "--fqbn",
+          "arduino:avr:uno",
+          "--warnings",
+          "all",
+          "--dry-run",
+          sketch_path_ino
+        )
+      end
+
+      it "Does not use --dry-run flag for arduino_cli version >= 0.14.0" do
```
[the remark this answers is not in the document]
Comment (hel06492) [reply spec/arduino_backend_spec.rb]: Done

**Comment by hel06492** [comment]
Hi @jgfoster @ianfixes 
Thanks for the suggestions. I've made changes according to those feedbacks.

**Comment by ianfixes** [comment]
@hel06492 Have you selected "[allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)" on your PR?  I can probably take care of the remaining issues.

**Comment by hel06492** [comment]
@ianfixes yes, it's already checked

**Comment by jgfoster** [comment]
@ianfixes, I'm going through the open PRs and wonder if this can be merged. Thanks!

**Comment by ianfixes** [comment]
I've merged this with a larger body of work in #334 -- thanks!

**Comment by ianfixes** [comment]
This has merged as part of #334 and was released in [`v1.4.0`](https://github.com/Arduino-CI/arduino_ci/releases/tag/v1.4.0)

**Pull request closed without merging**